pnp/PnpHelper.cs

```diff
@@ -0,0 +1,123 @@
+        }
+
+        // A writeable property is one which can be updated by an external source, eg. the service application, etc.
+        public static string CreateWriteableReportedPropertyPatch(
+            string propertyName,
+            string serializedPropertyValue,
+            int ackCode,
+            long ackVersion,
+            string serializedAckDescription = default,
+            string componentName = default)
+        {
+            string jsonString = string.IsNullOrWhiteSpace(componentName)
+                ?
+                    $"{{" +
+                    $"  \"{propertyName}\": " +
+                    $"      {{ " +
+                    $"          \"value\" : {serializedPropertyValue}," +
+                    $"          \"ac\" : {ackCode}, " +
+                    $"          \"av\" : {ackVersion}, " +
+                    $"          {(!string.IsNullOrWhiteSpace(serializedAckDescription) ? $"\"ad\": {serializedAckDescription}" : "")}" +
+                    $"      }} " +
+                    $"}}"
+                :
+                    $"{{" +
+                    $"  \"{componentName}\": " +
+                    $"      {{" +
+                    $"          \"{PropertyComponentIdentifierKey}\": \"{PropertyComponentIdentifierValue}\"," +
+                    $"          \"{propertyName}\": " +
+                    $"              {{ " +
+                    $"                  \"value\" : {serializedPropertyValue}," +
+                    $"                  \"ac\" : {ackCode}, " +
+                    $"                  \"av\" : {ackVersion}, " +
+                    $"                  {(!string.IsNullOrWhiteSpace(serializedAckDescription) ? $"\"ad\": {serializedAckDescription}" : "")}" +
+                    $"              }} " +
+                    $"      }} " +
+                    $"}}";
+            return jsonString.RemoveWhitespace();
+        }
+
+        public static (bool, T) GetPropertyFromTwin<T>(TwinCollection collection, string propertyName, string componentName = null)
+        {
+            // If the desired property update is for a root component, verify that property patch received contains the desired property update.
+            if (string.IsNullOrWhiteSpace(componentName))
+            {
+                return collection.Contains(propertyName) ? (true, (T)collection[propertyName]) : (false, default);
+            }
+
+            // If the desired property update is for a nested component, verify that the property patch received contains the pnp component identifier ("__t": "c"),
+            // and also the desired property update.
+            if (collection.Contains(componentName))
+            {
+                JObject componentProperty = collection[componentName];
+                if (componentProperty.ContainsKey(PropertyComponentIdentifierKey) && componentProperty.Value<string>(PropertyComponentIdentifierKey).Equals(PropertyComponentIdentifierValue))
```

Comment:
dont check for the __t flag. Desired Property Updates might not include the __t flag. See #7574815 for more details.